src/scheduler/MpiWorld.cpp

```diff
@@ -25,40 +30,109 @@ MpiWorld::MpiWorld()
   , cartProcsPerDim(2)
 {}
 
-faabric::scheduler::FunctionCallClient& MpiWorld::getFunctionCallClient(
-  const std::string& otherHost)
+void MpiWorld::sendRemoteMpiMessage(
+  int sendRank,
+  int recvRank,
+  const std::shared_ptr<faabric::MPIMessage>& msg)
 {
-    auto it = functionCallClients.find(otherHost);
-    if (it == functionCallClients.end()) {
-        // The second argument is forwarded to the client's constructor
-        auto _it = functionCallClients.try_emplace(otherHost, otherHost);
-        if (!_it.second) {
-            throw std::runtime_error("Error inserting remote endpoint");
+    // Assert the ranks are sane
+    assert(0 <= sendRank && sendRank < size);
+    assert(0 <= recvRank && recvRank < size);
+
+    // Initialise the endpoint vector if not initialised
```

Comment:
It looks like there's a few places where we're doing lazy initialisation in this class that could be captured in a single method, e.g. `initialiseAsyncForRank()`. This would be invoked call at the start of each `isend`/ `irecv` calls and do nothing if the initialisation had already happened, e.g. 

```
static thread_local asyncInitialised = false;

void initialiseAsyncForRank() {
    if(asyncInitialised) {
        return;
    }
    
    mpiMessageEndpoints = std::list(size * size, nullptr);
    
    // Other init
    
    asyncInitialised = true;
}
```